File: elixir/all-your-base/lib/all_your_base.ex
defmodule AllYourBase do

  defp base_to_num(list, base) do
    list
    |> Stream.zip(length(list)-1..0)
    |> Stream.map(fn {weight, exp} -> weight * base ** exp end)
    |> Enum.sum
  end

  defp num_to_base(num, base), do: num_to_base_acc(num, base, [])
  defp num_to_base_acc(0, _, []), do: [0]
  defp num_to_base_acc(0, _, list), do: list
  defp num_to_base_acc(num, base, list),
    do: num_to_base_acc(div(num, base), base, [rem(num, base) | list])



  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """
  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(_, _, output_base) when output_base < 2,
    do: {:error, "output base must be >= 2"}
  def convert(_, input_base, _) when input_base < 2,
    do: {:error, "input base must be >= 2"}
  def convert(digits, input_base, output_base) do
    cond do
      Enum.any?(digits, &(&1 < 0 or &1 >= input_base)) ->
        {:error, "all digits must be >= 0 and < input base"}
      true ->
        {:ok, digits |> base_to_num(input_base) |> num_to_base(output_base)}
    end
  end
end

File: c/all-your-base/all_your_base.c
#include "all_your_base.h"

size_t rebase(int8_t *digits, const int16_t input_base,
              const int16_t output_base, const size_t input_length) {
  if (input_length == 0) {
    return 0;
  }
  if (input_base < 2 || output_base < 2) {
    return 0;
  }
  size_t dec = 0;
  for (size_t i = 0; i < input_length; ++i) {
    if (digits[i] < 0 || digits[i] >= input_base) {
      return 0;
    }
    dec = dec * input_base + digits[i];
  }
  size_t size = 0;
  do {
    digits[size++] = dec % output_base;
    dec /= output_base;
  } while (dec > 0);
  for (size_t i = 0; i < size / 2; ++i) {
    digits[i] ^= digits[size - i - 1];
    digits[size - i - 1] ^= digits[i];
    digits[i] ^= digits[size - i - 1];
  }
  return size;
}
